### PR TITLE
global: general improvements

### PR DESCRIPTION
**Pull request by drjova**
~~**DO NOT** merge yet~~

### CODE REVIEW

**Comment by drjova** [comment]
@pamfilos @ioannistsanaktsidis can you take it for a spin?

It's now possible to create your custom data preparation function on your invenio app

```js
       $provide.decorator('invenioRecordsDirective', function($delegate) {
            var directive = $delegate[0];
            var link = directive.link;
            directive.compile = function() {
              return function Link(scope, element, attrs, ctrls) {
                ctrls.requestPrepare = function(url, method, model, extraParams) {
                  var _model = ctrls.removeEmptyValues(model);
                  return {
                    url: url,
                    method: method,
                    headers: extraParams.headers,
                    data: {
                      metadata: _model,
                    }
                  };
                }
                return link.apply(this, arguments);
              };
            };
            return $delegate;
          });
```

File: src/invenio-records-js/controllers/InvenioRecordsCtrl.js [hunk 1]
```diff
@@ -154,79 +154,116 @@ function InvenioRecordsCtrl($scope, $rootScope, $q, $window, $location,
...
+    // Pass all the data information
+    var _data = angular.merge(
+      {},
+      extraParams.data || {},
+      _model
+    );
+    return {
+      url: url,
+      method: (method || 'PUT').toUpperCase(),
+      data: _data,
```
Comment (jirikuncar) [inline src/invenio-records-js/controllers/InvenioRecordsCtrl.js]: 👍

**Comment by drjova** [comment]
The PR updated with a more DRY approach for different use cases, it's possible to change the data and clean logic with a ``service`` decorator:


```js
.config([ '$provide', function($provide) {

  $provide.decorator('InvenioRecordsAPI', [
    '$delegate',
    function myServiceDecorator($delegate) {

      // keep the existing function
      var getData = $delegate.getData;
     
      function newCleanData(data, unwanted) {
        // unicorns and stuff
        return data;
      }

      function newGetData(model, extraParams, endpoints) {
        var _model = {
            metadata: model
        };
        // More unicorns and stuff

        // Pass the arguments to the existing function without re-writting the code
        getData.apply($delegate, _model, extraParams, endpoints);
      }

      // Override the cleanData
      $delegate.cleanData = newCleanData;

      // Override the getData
      $delegate.getData = newGetData;

      // return the delegate
      return $delegate;
    }
  ]);
}]);
```

**Comment by ioannistsanaktsidis** [comment]
This one works for me and @pamfilos. I think it can go on. 
cc @jirikuncar , @lnielsen

**Pull request merged**